ldrshared/clients/messenger/pubsub.py

```diff
@@ -0,0 +1,250 @@
+class PubSubMessage(BaseMessage):
+    message: Message
+
+    def get_attribute(self, key: str, default=None):
+        return self.message.attributes.get(key, default)
+
+    def get(self, key=None, default=None) -> Union[dict, str, list]:
+        data: str = self.message.data.decode("utf-8")
+        if key:
+            data_as_dict: dict = json.loads(data)
+            return data_as_dict.get(key, default)
+
+        try:
+            return json.loads(data)
+        except:
+            return data
+
+
+# TODO - will need to be implemented with an actual database
+# before we start subscribing from more than one location
+class Deduplicator:
+    """
+    Uses simple database operations to check if a given message has been
+    received before.
+    """
+
+    def __init__(self, dbname: str, message_retention: dict):
+        self.db = TinyDB(dbname)
+        self.message_retention: datetime.timedelta = datetime.timedelta(
+            days=message_retention["days"] if "days" in message_retention else 0,
+            hours=message_retention["hours"] if "hours" in message_retention else 0,
+            minutes=message_retention["minutes"]
+            if "minutes" in message_retention
+            else 0,
+            seconds=message_retention["seconds"]
+            if "seconds" in message_retention
+            else 0,
+        )
+
+    @staticmethod
+    def _message_to_dict(message: Message) -> dict:
+        """
+        Convert type pubsub_v1.subscriber.message.Message to a simple dict
+        """
+        return {
+            "content": message.data.decode("utf-8"),
+            "time_stamp": message.attributes.get(TIME_STAMP_KEY),
+        }
+
+    def store(self, message: Message):
+        """
+        Stores the content and timestamp of a single message
+        """
+        self.db.insert(self._message_to_dict(message))
+        self._housekeeping()
+
+    def is_new_message(self, message: Message) -> bool:
+        """
+        Check if the message received is a new message
+        """
+        new_message_as_dict = self._message_to_dict(message)
+        q = Query()
+        found = self.db.search(
+            (q.content == new_message_as_dict["content"])
+            & (q.time_stamp == new_message_as_dict["time_stamp"])
+        )
+        return len(found) == 0
+
+    def message_is_in_list(
+        self, message: Message, list_of_messages: List[Message]
+    ) -> bool:
+        """
+        Does the received message already exist in the provided list of messages.
+
+        Note, we need to cast the message to a much simplified dict form as pubsub.Message
+        attributes otherwise render all messages unique, even where they are duplicates
+        in term of being the same pushed message.
+        """
+        new_message = self._message_to_dict(message)
+        known_messages = [self._message_to_dict(x) for x in list_of_messages]
+
+        return new_message in known_messages
```

Comment:
It'd probably be more efficient to have `known_messages` instantiated as a set rather than a list. But there may not be much in it.